drivers/adc/adc_stm32.c

```diff
@@ -422,51 +422,46 @@ static int start_read(const struct device *dev,
 	defined(CONFIG_SOC_SERIES_STM32L4X) || \
 	defined(CONFIG_SOC_SERIES_STM32WBX) || \
 	defined(CONFIG_SOC_SERIES_STM32WLX)
-	switch (sequence->oversampling) {
-	case 0:
+
+static const uint32_t table_oversampling_shift[] = {
+	0,
+	LL_ADC_OVS_SHIFT_RIGHT_1,
+	LL_ADC_OVS_SHIFT_RIGHT_2,
+	LL_ADC_OVS_SHIFT_RIGHT_3,
+	LL_ADC_OVS_SHIFT_RIGHT_4,
+	LL_ADC_OVS_SHIFT_RIGHT_5,
+	LL_ADC_OVS_SHIFT_RIGHT_6,
+	LL_ADC_OVS_SHIFT_RIGHT_7,
+	LL_ADC_OVS_SHIFT_RIGHT_8,
+};
+
+	uint32_t oversampling_ratio = sequence->oversampling;
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && defined(ADC_VER_V5_V90)
+	if (adc != ADC3) {
+		oversampling_ratio = 1<<sequence->oversampling;
+	}
+#endif
+	if (sequence->oversampling == 0) {
 		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_DISABLE);
-		break;
-	case 1:
-		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_GRP_REGULAR_CONTINUED);
-		LL_ADC_ConfigOverSamplingRatioShift(adc, LL_ADC_OVS_RATIO_2,
-						    LL_ADC_OVS_SHIFT_RIGHT_1);
-		break;
-	case 2:
-		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_GRP_REGULAR_CONTINUED);
-		LL_ADC_ConfigOverSamplingRatioShift(adc, LL_ADC_OVS_RATIO_4,
-						    LL_ADC_OVS_SHIFT_RIGHT_2);
-		break;
-	case 3:
-		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_GRP_REGULAR_CONTINUED);
-		LL_ADC_ConfigOverSamplingRatioShift(adc, LL_ADC_OVS_RATIO_8,
-						    LL_ADC_OVS_SHIFT_RIGHT_3);
-		break;
-	case 4:
-		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_GRP_REGULAR_CONTINUED);
-		LL_ADC_ConfigOverSamplingRatioShift(adc, LL_ADC_OVS_RATIO_16,
-						    LL_ADC_OVS_SHIFT_RIGHT_4);
-		break;
-	case 5:
+	} else if (sequence->oversampling <= 8) {
 		LL_ADC_SetOverSamplingScope(adc, LL_ADC_OVS_GRP_REGULAR_CONTINUED);
```

Comment:
So this works for ADC3 also? At least the bits 10 and 9 are the same in cfg2 register for adc1/2/3, so probably yes.

Reply:
Yes, the settings are in the same place in `CFG2`.